### ISSUE TITLE
History: sync_on_enter: false doesn't record history at all

### ISSUE THREAD

**Issue by maxim-uvarov** [nushell/nushell#7790]
### Describe the bug

With such settings history is not recorded. 
```
  history: {
    max_size: 30000 # Session has to be reloaded for this to take effect
    sync_on_enter: false # Enable to share history between multiple sessions, else you have to close the session to write history to file
    file_format: "plaintext" # "sqlite" or "plaintext"
  }
```

### How to reproduce

Set history setting `sync_on_enter: false` 
Open terminal
Open nu
enter command
exit nu
open nu
check if the previous command is in history

### Expected behavior

It would be nice, if I could use history from the current session

### Screenshots

https://user-images.githubusercontent.com/4896754/213190877-098aaa10-bdb3-457c-8075-35d6fe3d2142.mp4

### Configuration

| key                | value                                                             |
| ------------------ | ----------------------------------------------------------------- |
| version            | 0.74.0                                                            |
| branch             |                                                                   |
| commit_hash        |                                                                   |
| build_os           | macos-aarch64                                                     |
| build_target       | aarch64-apple-darwin                                              |
| rust_version       | rustc 1.66.0 (69f9c33d7 2022-12-12) (built from a source tarball) |
| cargo_version      | cargo 1.66.0 (d65d197ad 2022-11-15)                               |
| pkg_version        | 0.74.0                                                            |
| build_time         | 2023-01-10 20:12:41 +00:00                                        |
| build_rust_channel | release                                                           |
| features           | database, default, trash, which, zip                              |
| installed_plugins  |                                                                   |

**Comment by sholderbach** [comment]
Mhh struggling to reproduce it locally (with `plaintext` and `sync_on_enter: false`). New entries are added to the existing history file. You use your local file system for the nushell configuration directory or did you wipe the history file when changing the configuration?

If we can narrow this problem down it might be an issue in the [`reedline`](https://github.com/nushell/reedline) `FileBackedHistory`.

**Comment by fdncred** [comment]
@maxim-uvarov in your example video, what happens if you do these two commands after you've typed some other commands?
1. open $nu.history-path
2. history
Just a theory, but just because you don't have hints doesn't necessarily mean the history isn't being recorded. There could be a bug somewhere else.

**Comment by maxim-uvarov** [comment]
> what happens if you do these two commands after you've typed some other commands?
> open $nu.history-path

It gives nothing 

> history

It give nothing too

<img width="385" alt="image" src="https://user-images.githubusercontent.com/4896754/213366716-58d67827-2fa2-4d3a-9f71-0463ad0a9488.png">

I recorded another quick video, that might give you some cue of what is wrong with my setup.  
In the video I removed history.txt file. Several times restarted nu. And history is not recorded at all. 

Also I attached my config.nu (config-20230119.txt) file. 

I use starship, and also I made the folder '/Users/user/Library/Application Support/nushell' a git repository. Might they be a problem? 

And for information: If I set up "sqlite" storage - history works as expected, though the setting `sync_on_enter: false` doesn't make any changes in behaviour: if I start a new session of nu, the older sessions would see all the commands in history from the newly started session.

https://user-images.githubusercontent.com/4896754/213366982-be68e072-1702-4203-a0b7-b5b91db544c6.mp4


[config-20230119.txt](https://github.com/nushell/nushell/files/10453947/config-20230119.txt)

**Comment by fdncred** [comment]
Just to be sure, what is the output of this command?
```
$env.config.history
```

**Comment by maxim-uvarov** [comment]
<img width="340" alt="image" src="https://user-images.githubusercontent.com/4896754/213469059-2409feac-86b1-4df6-961a-2018978a1506.png">

**Comment by maxim-uvarov** [comment]
> $env.config.history

that is handy! I didn't know about this! 🙏🏼️

**Comment by fdncred** [comment]
I really have no idea what is going on here and why you're not getting a complete history.txt file. You'd think that if there were problems, errors would be thrown.

**Comment by maxim-uvarov** [comment]
I can definetly live with the "sqlite" setting. If I ever find out, what was the problem - I'll add it here.

**Comment by maxim-uvarov** [comment]
After some time history appeared working again. And now works as expected, as far as i can see. 

<img width="459" alt="image" src="https://user-images.githubusercontent.com/4896754/213478187-e17a36f6-ce8a-4de8-8338-44bcd9449477.png">

**Comment by maxim-uvarov** [comment]
sorry. In the previous comment I was wrong: I forgot that we are testing "sync_on_enter":false behaviour. And with it my history doesn't work.

**Comment by meskill** [comment]
Same issue here.

With option `sync_on_enter: false` the history is only saved after exiting nushell with hotkey `CTRL + D` . If shell is closed by `exit` command or by closing the whole terminal nothing is saved

**Comment by sholderbach** [comment]
> With option `sync_on_enter: false` the history is only saved after exiting nushell with hotkey `CTRL + D` . If shell is closed by `exit` command or by closing the whole terminal nothing is saved

Great sleuthing! That is probably why I wasn't noticing it as my muscle memory uses `^D` basically all the time!

**Comment by ineu** [comment]
`C-d` is fine most of the time, that's what I use as well when dealing with a single shell. However, it becomes a real issue when used in tmux+tmuxinator or zellij, that is: when you have like 10 multiplexer windows running different programs. To exit the session you would have to type `C-d` 10 times to close all shells one by one, sometimes combined with `C-c` to kill programs running in windows.

**Comment by Ecorous** [comment]
Cannot reproduce this on 0.106.1 with the config file 
```
$env.config.history = {
  max_size: 30000 # Session has to be reloaded for this to take effect
  sync_on_enter: false # Enable to share history between multiple sessions, else you have to close the session to write history to file
  file_format: "plaintext" # "sqlite" or "plaintext"
}
```
and a clean `history.txt`. 
History seems to be saved fine when using `exit`

**Comment by maxim-uvarov** [comment]
Yep, now it clearly states that sqlite format is needed. @Ecorous thanks for checking

![Image](https://github.com/user-attachments/assets/674c0729-ce28-4057-8992-a680b85ca799)

I guess that is guess we can close the issue

**Comment by weirdan** [comment]
@maxim-uvarov it looks like it complains about `isolation = true`, not about `sync_on_enter`
